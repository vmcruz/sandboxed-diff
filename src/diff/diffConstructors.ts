import { type DiffResult, DiffConstructorArgs } from '../types';
import { ChangeType } from '../utils/constants';
import { emptyShellClone, isObject } from '../utils/fns';

function diffConstructors({
  recursiveDiff,
  lhs,
  rhs,
  config,
  depth,
  parent,
  seen,
  initialChangeType,
  path,
  startedAt,
}: DiffConstructorArgs): DiffResult[] {
  // Different Constructors
  let modLhs = lhs;
  let modRhs = rhs;
  let defaultChangeType = initialChangeType;

  if (isObject(rhs) || rhs?.constructor) {
    // Added in rhs. Use rhs prototype to traverse the added properties
    modLhs = emptyShellClone(rhs);

    defaultChangeType = ChangeType.ADD;
  } else if (isObject(lhs) || lhs?.constructor) {
    // Removed in rhs. Use lhs prototype to traverse the removed properties
    modRhs = emptyShellClone(lhs);

    defaultChangeType = ChangeType.REMOVE;
  } else {
    // Fail for unknown edge cases
    throw new Error("Edge case raised, I don't know how to handle this input");
  }

  return recursiveDiff({
    lhs: modLhs,
    rhs: modRhs,
    config,
    depth,
    parent,
    seen,
    initialChangeType: defaultChangeType,
    path,
    startedAt,
  });
}

export default diffConstructors;
