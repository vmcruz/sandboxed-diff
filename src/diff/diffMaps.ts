import { DiffResult, DiffMapArgs } from '../types';
import { ITERATION_TIMEOUT_CHECK } from '../utils/constants';
import { isIterable } from '../utils/fns';
import {
  buildResult,
  getObjectChangeResult,
  getPathHint,
  maxKeysSecurityCheck,
  timeoutSecurityCheck,
} from './shared';

function diffMaps({
  recursiveDiff,
  lhs,
  rhs,
  config,
  depth,
  parent,
  seen,
  initialChangeType,
  path,
  startedAt,
}: DiffMapArgs): DiffResult[] {
  const result: DiffResult[] = [];

  maxKeysSecurityCheck(lhs.size, config);
  maxKeysSecurityCheck(rhs.size, config);

  const mergedMapKeys = new Set([...lhs.keys(), ...rhs.keys()]);

  let i = 0;
  for (const key of mergedMapKeys) {
    if (i++ % ITERATION_TIMEOUT_CHECK === 0) {
      timeoutSecurityCheck(startedAt, config);
    }

    const keyInLhs = lhs.has(key);
    const keyInRhs = rhs.has(key);
    const lhsValue = keyInLhs ? lhs.get(key) : null;
    const rhsValue = keyInRhs ? rhs.get(key) : null;
    const hint = getPathHint(config, 'map');
    const pathUpdate = hint ? [hint, key] : [key];
    const updatedPath = [...path, ...pathUpdate];

    if (isIterable(lhsValue) || isIterable(rhsValue)) {
      result.push(
        ...recursiveDiff({
          lhs: lhsValue,
          rhs: rhsValue,
          config,
          depth: depth + 1,
          parent: key,
          seen,
          initialChangeType,
          path: updatedPath,
          startedAt,
        })
      );
      continue;
    }

    result.push(
      ...getObjectChangeResult(lhs, rhs, depth, key, key, config, updatedPath)
    );
  }

  return buildResult(
    rhs,
    result,
    depth,
    initialChangeType,
    parent,
    path,
    config
  );
}

export default diffMaps;
