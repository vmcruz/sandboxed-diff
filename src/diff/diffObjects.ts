import { type DiffResult, DiffObjectArgs } from '../types';
import { ITERATION_TIMEOUT_CHECK } from '../utils/constants';
import { getEnumerableKeys, isIterable } from '../utils/fns';
import {
  buildResult,
  getObjectChangeResult,
  maxKeysSecurityCheck,
  shouldRedactValue,
  timeoutSecurityCheck,
} from './shared';

function diffObjects({
  recursiveDiff,
  lhs,
  rhs,
  config,
  depth,
  parent,
  seen,
  initialChangeType,
  path,
  startedAt,
}: DiffObjectArgs): DiffResult[] {
  const result: DiffResult[] = [];

  const lhsKeys = getEnumerableKeys(lhs);
  maxKeysSecurityCheck(lhsKeys.length, config);

  const rhsKeys = getEnumerableKeys(rhs);
  maxKeysSecurityCheck(rhsKeys.length, config);

  const keys = new Set<string | number | symbol>([...lhsKeys, ...rhsKeys]);

  let i = 0;
  for (const key of keys) {
    if (i++ % ITERATION_TIMEOUT_CHECK === 0) {
      timeoutSecurityCheck(startedAt, config);
    }

    const redactable = shouldRedactValue(key, config);
    const lhsValue = Array.isArray(lhs) ? lhs[key as number] : lhs?.[key];
    const rhsValue = Array.isArray(rhs) ? rhs[key as number] : rhs?.[key];
    const numericKey = typeof key !== 'symbol' ? Number(key) : NaN;
    const parsedKey = isNaN(numericKey) ? key : numericKey;
    const updatedPath = [...path, parsedKey];

    if (!redactable && (isIterable(lhsValue) || isIterable(rhsValue))) {
      result.push(
        ...recursiveDiff({
          lhs: lhsValue,
          rhs: rhsValue,
          config,
          depth: depth + 1,
          parent: parsedKey,
          seen,
          initialChangeType,
          path: updatedPath,
          startedAt,
        })
      );
      continue;
    }

    result.push(
      ...getObjectChangeResult(
        lhs,
        rhs,
        depth,
        key,
        parsedKey,
        config,
        updatedPath
      )
    );
  }

  seen.delete(lhs);
  seen.delete(rhs);

  return buildResult(
    rhs,
    result,
    depth,
    initialChangeType,
    parent,
    path,
    config
  );
}

export default diffObjects;
