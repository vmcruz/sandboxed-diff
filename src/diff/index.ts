import { type DiffResult, RecursiveDiffArgs } from '../types';
import diffObjects from './diffObjects';
import diffSets from './diffSets';
import diffMaps from './diffMaps';
import diffConstructors from './diffConstructors';
import { ChangeType } from '../utils/constants';
import { areObjects, isNullOrUndefined, isPrimitive } from '../utils/fns';
import {
  createReplacer,
  includeDiffType,
  lastPathValue,
  timeoutSecurityCheck,
  stringify,
} from './shared';

function recursiveDiff({
  lhs,
  rhs,
  config,
  depth,
  parent,
  seen,
  initialChangeType,
  path,
  startedAt,
}: RecursiveDiffArgs): DiffResult[] {
  // Prevents stack overflow
  if (depth > config.maxDepth) {
    throw new Error('Max depth exceeded!');
  }

  // Checks for timeout before proceeding
  timeoutSecurityCheck(startedAt, config);

  // Allows printing the first level of circular references
  const lhsSeen = seen.get(lhs) ?? 0;
  const rhsSeen = seen.get(rhs) ?? 0;

  if (lhsSeen > 1 || rhsSeen > 1) {
    if (!includeDiffType(initialChangeType, config)) return [];

    return [
      {
        type: initialChangeType,
        str: `${stringify(parent, config)}: [Circular],`,
        depth: depth - 1,
        path,
      },
    ];
  }

  // Prevents circular references except for primitives
  if (typeof lhs === 'object' && lhs !== null) seen.set(lhs, lhsSeen + 1);
  if (typeof rhs === 'object' && rhs !== null) seen.set(rhs, rhsSeen + 1);

  const args = {
    recursiveDiff,
    lhs,
    rhs,
    config,
    depth,
    parent,
    seen,
    initialChangeType,
    path,
    startedAt,
  };

  // Handles comparing primitives
  if (isPrimitive(lhs) && isPrimitive(rhs)) {
    // Return early if it's the same primitive
    if (config.strict ? lhs === rhs : lhs == rhs) {
      return [];
    }

    const parentDepth = depth - 1;
    const lhsValue = JSON.stringify(lhs, createReplacer(config));
    const rhsValue = JSON.stringify(rhs, createReplacer(config));

    const result = [];

    if (includeDiffType(ChangeType.UPDATE, config)) {
      if (!config.showUpdatedOnly) {
        result.push({
          type: ChangeType.REMOVE,
          str: `${lhsValue}`,
          depth: parentDepth,
          path: [...path, lastPathValue(ChangeType.REMOVE, lhs)],
        });
      }

      result.push({
        type: ChangeType.UPDATE,
        str: `${rhsValue}`,
        depth: parentDepth,
        path: [...path, lastPathValue(ChangeType.UPDATE, rhs)],
      });
    }

    return result;
  }

  // Different constructor handler
  if (
    !areObjects(lhs, rhs) && // Skips for Object.create(null) vs {}
    (lhs?.constructor !== rhs?.constructor ||
      (isNullOrUndefined(lhs) && rhs) ||
      (lhs && isNullOrUndefined(rhs)))
  ) {
    return diffConstructors(args);
  }

  // SETS
  if (lhs instanceof Set && rhs instanceof Set) {
    return diffSets(args);
  }

  // MAPS
  if (lhs instanceof Map && rhs instanceof Map) {
    return diffMaps(args);
  }

  // Arrays and Objects
  return diffObjects(args);
}

export default recursiveDiff;
