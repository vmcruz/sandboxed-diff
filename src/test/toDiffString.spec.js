import diff, { ChangeType } from '..';

const complexKey = { complex: 'key' }; // Shared reference

const a = {
  id: 123,
  name: 'Alice',
  email: 'alice@example.com',
  password: 'secret',
  preferences: {
    theme: 'dark',
    notifications: true,
    layout: ['grid', 'compact'],
  },
  metadata: new Map([
    [complexKey, 'value'], // Complex key (shared reference)
    ['role', 'admin'],
    ['permissions', new Set(['read', 'write', 'execute'])],
  ]),
  tags: new Set(['user', 'active']),
  lastLogin: new Date('2024-03-01T12:00:00Z').getTime(),
  custom: Object.create(null, { key: { value: 'value', enumerable: true } }),

  // Additional edge cases:
  emptyArray: [],
  emptyObject: {},
  nestedEmpty: { a: {} },
};

// Circular reference
a.self = a;

const b = {
  id: 123, // Unchanged
  name: 'Alice Johnson', // Changed
  email: 'alice@example.com', // Unchanged
  password: 'newsecret', // Should be redacted
  preferences: {
    theme: 'light', // Changed
    notifications: false, // Changed
    layout: ['list'], // Changed
    newFeature: true, // **Added property**
  },
  metadata: new Map([
    [complexKey, 'newValue'], // **Value changed, key remains the same**
    ['role', 'user'], // Changed
    ['permissions', new Set(['read', 'execute'])], // Removed "write"
    ['createdAt', new Date('2024-03-01T12:00:00Z').getTime()], // **Added property**
  ]),
  tags: new Set(['user', 'inactive', 'premium']), // **Added "premium" tag**
  lastLogin: new Date('2024-03-22T08:30:00Z').getTime(), // Changed
  custom: Object.create(null, { key: { value: 'modified', enumerable: true } }), // Changed in Object.create(null)

  // Additional edge cases:
  emptyArray: ['unexpected'], // Previously empty
  emptyObject: { newKey: 'new' }, // **Added property**
  nestedEmpty: { a: { b: 'value' } }, // Previously empty nested object
  newProp: 'New Value', // **New property at root level**
};

// Circular reference
b.self = b;

describe('toDiffString', () => {
  it('outputs with defaults', () => {
    const result = diff(a, b);

    expect(result.toDiffString()).toMatchSnapshot();
  });

  it('outputs without color', () => {
    const result = diff(a, b);

    const config = {
      withColors: false,
    };

    expect(result.toDiffString(config)).toMatchSnapshot();
  });

  it('outputs custom colors and symbols', () => {
    const result = diff(a, b);

    const config = {
      colors: {
        [ChangeType.NOOP]: (str) => `\x1b[36m${str}\x1b[0m`,
      },
      symbols: {
        [ChangeType.ADD]: '@',
        [ChangeType.REMOVE]: '#',
        [ChangeType.UPDATE]: '$',
        [ChangeType.NOOP]: '%',
      },
    };

    expect(result.toDiffString(config)).toMatchSnapshot();
  });

  it('outputs with custom wrapper and indentSize', () => {
    const result = diff(a, b);

    const config = {
      wrapper: ['```diff', '```'],
      indentSize: 5,
    };

    expect(result.toDiffString(config)).toMatchSnapshot();
  });

  it('outputs empty wrapper strings if missing', () => {
    const result = diff(a, b);

    expect(result.toDiffString({ wrapper: [] })).toMatchSnapshot();
    expect(result.toDiffString({ wrapper: null })).toMatchSnapshot();
  });
});
