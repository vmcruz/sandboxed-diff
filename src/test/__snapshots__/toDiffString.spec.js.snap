// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toDiffString outputs custom colors and symbols 1`] = `
"[36m%{[0m
[36m%  "id": 123,[0m
[31m#  "name": "Alice",[0m
[33m$  "name": "Alice Johnson",[0m
[36m%  "email": "alice@example.com",[0m
[31m#  "password": "*****",[0m
[33m$  "password": "*****",[0m
[36m%  "preferences": {[0m
[31m#    "theme": "dark",[0m
[33m$    "theme": "light",[0m
[31m#    "notifications": true,[0m
[33m$    "notifications": false,[0m
[36m%    "layout": [[0m
[31m#      0: "grid",[0m
[33m$      0: "list",[0m
[31m#      1: "compact",[0m
[36m%    ],[0m
[32m@    "newFeature": true,[0m
[36m%  },[0m
[36m%  "metadata": Map (4) {[0m
[31m#    {"complex":"key"}: "value",[0m
[33m$    {"complex":"key"}: "newValue",[0m
[31m#    "role": "admin",[0m
[33m$    "role": "user",[0m
[36m%    "permissions": Set [[0m
[36m%      "read",[0m
[31m#      "write",[0m
[36m%      "execute",[0m
[36m%    ],[0m
[32m@    "createdAt": 1709294400000,[0m
[36m%  },[0m
[36m%  "tags": Set [[0m
[36m%    "user",[0m
[31m#    "active",[0m
[32m@    "inactive",[0m
[32m@    "premium",[0m
[36m%  ],[0m
[31m#  "lastLogin": 1709294400000,[0m
[33m$  "lastLogin": 1711096200000,[0m
[36m%  "custom": {[0m
[31m#    "key": "value",[0m
[33m$    "key": "modified",[0m
[36m%  },[0m
[36m%  "emptyArray": [[0m
[32m@    0: "unexpected",[0m
[36m%  ],[0m
[36m%  "emptyObject": {[0m
[32m@    "newKey": "new",[0m
[36m%  },[0m
[36m%  "nestedEmpty": {[0m
[36m%    "a": {[0m
[32m@      "b": "value",[0m
[36m%    },[0m
[36m%  },[0m
[36m%  "self": {[0m
[36m%    "id": 123,[0m
[31m#    "name": "Alice",[0m
[33m$    "name": "Alice Johnson",[0m
[36m%    "email": "alice@example.com",[0m
[31m#    "password": "*****",[0m
[33m$    "password": "*****",[0m
[36m%    "preferences": {[0m
[31m#      "theme": "dark",[0m
[33m$      "theme": "light",[0m
[31m#      "notifications": true,[0m
[33m$      "notifications": false,[0m
[36m%      "layout": [[0m
[31m#        0: "grid",[0m
[33m$        0: "list",[0m
[31m#        1: "compact",[0m
[36m%      ],[0m
[32m@      "newFeature": true,[0m
[36m%    },[0m
[36m%    "metadata": Map (4) {[0m
[31m#      {"complex":"key"}: "value",[0m
[33m$      {"complex":"key"}: "newValue",[0m
[31m#      "role": "admin",[0m
[33m$      "role": "user",[0m
[36m%      "permissions": Set [[0m
[36m%        "read",[0m
[31m#        "write",[0m
[36m%        "execute",[0m
[36m%      ],[0m
[32m@      "createdAt": 1709294400000,[0m
[36m%    },[0m
[36m%    "tags": Set [[0m
[36m%      "user",[0m
[31m#      "active",[0m
[32m@      "inactive",[0m
[32m@      "premium",[0m
[36m%    ],[0m
[31m#    "lastLogin": 1709294400000,[0m
[33m$    "lastLogin": 1711096200000,[0m
[36m%    "custom": {[0m
[31m#      "key": "value",[0m
[33m$      "key": "modified",[0m
[36m%    },[0m
[36m%    "emptyArray": [[0m
[32m@      0: "unexpected",[0m
[36m%    ],[0m
[36m%    "emptyObject": {[0m
[32m@      "newKey": "new",[0m
[36m%    },[0m
[36m%    "nestedEmpty": {[0m
[36m%      "a": {[0m
[32m@        "b": "value",[0m
[36m%      },[0m
[36m%    },[0m
[36m%    "self": [Circular],[0m
[32m@    "newProp": "New Value",[0m
[36m%  },[0m
[32m@  "newProp": "New Value",[0m
[36m%}[0m"
`;

exports[`toDiffString outputs empty wrapper strings if missing 1`] = `
"{[0m
   "id": 123,[0m
[31m-  "name": "Alice",[0m
[33m!  "name": "Alice Johnson",[0m
   "email": "alice@example.com",[0m
[31m-  "password": "*****",[0m
[33m!  "password": "*****",[0m
   "preferences": {[0m
[31m-    "theme": "dark",[0m
[33m!    "theme": "light",[0m
[31m-    "notifications": true,[0m
[33m!    "notifications": false,[0m
     "layout": [[0m
[31m-      0: "grid",[0m
[33m!      0: "list",[0m
[31m-      1: "compact",[0m
     ],[0m
[32m+    "newFeature": true,[0m
   },[0m
   "metadata": Map (4) {[0m
[31m-    {"complex":"key"}: "value",[0m
[33m!    {"complex":"key"}: "newValue",[0m
[31m-    "role": "admin",[0m
[33m!    "role": "user",[0m
     "permissions": Set [[0m
       "read",[0m
[31m-      "write",[0m
       "execute",[0m
     ],[0m
[32m+    "createdAt": 1709294400000,[0m
   },[0m
   "tags": Set [[0m
     "user",[0m
[31m-    "active",[0m
[32m+    "inactive",[0m
[32m+    "premium",[0m
   ],[0m
[31m-  "lastLogin": 1709294400000,[0m
[33m!  "lastLogin": 1711096200000,[0m
   "custom": {[0m
[31m-    "key": "value",[0m
[33m!    "key": "modified",[0m
   },[0m
   "emptyArray": [[0m
[32m+    0: "unexpected",[0m
   ],[0m
   "emptyObject": {[0m
[32m+    "newKey": "new",[0m
   },[0m
   "nestedEmpty": {[0m
     "a": {[0m
[32m+      "b": "value",[0m
     },[0m
   },[0m
   "self": {[0m
     "id": 123,[0m
[31m-    "name": "Alice",[0m
[33m!    "name": "Alice Johnson",[0m
     "email": "alice@example.com",[0m
[31m-    "password": "*****",[0m
[33m!    "password": "*****",[0m
     "preferences": {[0m
[31m-      "theme": "dark",[0m
[33m!      "theme": "light",[0m
[31m-      "notifications": true,[0m
[33m!      "notifications": false,[0m
       "layout": [[0m
[31m-        0: "grid",[0m
[33m!        0: "list",[0m
[31m-        1: "compact",[0m
       ],[0m
[32m+      "newFeature": true,[0m
     },[0m
     "metadata": Map (4) {[0m
[31m-      {"complex":"key"}: "value",[0m
[33m!      {"complex":"key"}: "newValue",[0m
[31m-      "role": "admin",[0m
[33m!      "role": "user",[0m
       "permissions": Set [[0m
         "read",[0m
[31m-        "write",[0m
         "execute",[0m
       ],[0m
[32m+      "createdAt": 1709294400000,[0m
     },[0m
     "tags": Set [[0m
       "user",[0m
[31m-      "active",[0m
[32m+      "inactive",[0m
[32m+      "premium",[0m
     ],[0m
[31m-    "lastLogin": 1709294400000,[0m
[33m!    "lastLogin": 1711096200000,[0m
     "custom": {[0m
[31m-      "key": "value",[0m
[33m!      "key": "modified",[0m
     },[0m
     "emptyArray": [[0m
[32m+      0: "unexpected",[0m
     ],[0m
     "emptyObject": {[0m
[32m+      "newKey": "new",[0m
     },[0m
     "nestedEmpty": {[0m
       "a": {[0m
[32m+        "b": "value",[0m
       },[0m
     },[0m
     "self": [Circular],[0m
[32m+    "newProp": "New Value",[0m
   },[0m
[32m+  "newProp": "New Value",[0m
}[0m"
`;

exports[`toDiffString outputs empty wrapper strings if missing 2`] = `
"{[0m
   "id": 123,[0m
[31m-  "name": "Alice",[0m
[33m!  "name": "Alice Johnson",[0m
   "email": "alice@example.com",[0m
[31m-  "password": "*****",[0m
[33m!  "password": "*****",[0m
   "preferences": {[0m
[31m-    "theme": "dark",[0m
[33m!    "theme": "light",[0m
[31m-    "notifications": true,[0m
[33m!    "notifications": false,[0m
     "layout": [[0m
[31m-      0: "grid",[0m
[33m!      0: "list",[0m
[31m-      1: "compact",[0m
     ],[0m
[32m+    "newFeature": true,[0m
   },[0m
   "metadata": Map (4) {[0m
[31m-    {"complex":"key"}: "value",[0m
[33m!    {"complex":"key"}: "newValue",[0m
[31m-    "role": "admin",[0m
[33m!    "role": "user",[0m
     "permissions": Set [[0m
       "read",[0m
[31m-      "write",[0m
       "execute",[0m
     ],[0m
[32m+    "createdAt": 1709294400000,[0m
   },[0m
   "tags": Set [[0m
     "user",[0m
[31m-    "active",[0m
[32m+    "inactive",[0m
[32m+    "premium",[0m
   ],[0m
[31m-  "lastLogin": 1709294400000,[0m
[33m!  "lastLogin": 1711096200000,[0m
   "custom": {[0m
[31m-    "key": "value",[0m
[33m!    "key": "modified",[0m
   },[0m
   "emptyArray": [[0m
[32m+    0: "unexpected",[0m
   ],[0m
   "emptyObject": {[0m
[32m+    "newKey": "new",[0m
   },[0m
   "nestedEmpty": {[0m
     "a": {[0m
[32m+      "b": "value",[0m
     },[0m
   },[0m
   "self": {[0m
     "id": 123,[0m
[31m-    "name": "Alice",[0m
[33m!    "name": "Alice Johnson",[0m
     "email": "alice@example.com",[0m
[31m-    "password": "*****",[0m
[33m!    "password": "*****",[0m
     "preferences": {[0m
[31m-      "theme": "dark",[0m
[33m!      "theme": "light",[0m
[31m-      "notifications": true,[0m
[33m!      "notifications": false,[0m
       "layout": [[0m
[31m-        0: "grid",[0m
[33m!        0: "list",[0m
[31m-        1: "compact",[0m
       ],[0m
[32m+      "newFeature": true,[0m
     },[0m
     "metadata": Map (4) {[0m
[31m-      {"complex":"key"}: "value",[0m
[33m!      {"complex":"key"}: "newValue",[0m
[31m-      "role": "admin",[0m
[33m!      "role": "user",[0m
       "permissions": Set [[0m
         "read",[0m
[31m-        "write",[0m
         "execute",[0m
       ],[0m
[32m+      "createdAt": 1709294400000,[0m
     },[0m
     "tags": Set [[0m
       "user",[0m
[31m-      "active",[0m
[32m+      "inactive",[0m
[32m+      "premium",[0m
     ],[0m
[31m-    "lastLogin": 1709294400000,[0m
[33m!    "lastLogin": 1711096200000,[0m
     "custom": {[0m
[31m-      "key": "value",[0m
[33m!      "key": "modified",[0m
     },[0m
     "emptyArray": [[0m
[32m+      0: "unexpected",[0m
     ],[0m
     "emptyObject": {[0m
[32m+      "newKey": "new",[0m
     },[0m
     "nestedEmpty": {[0m
       "a": {[0m
[32m+        "b": "value",[0m
       },[0m
     },[0m
     "self": [Circular],[0m
[32m+    "newProp": "New Value",[0m
   },[0m
[32m+  "newProp": "New Value",[0m
}[0m"
`;

exports[`toDiffString outputs with custom wrapper and indentSize 1`] = `
"\`\`\`diff
{[0m
      "id": 123,[0m
[31m-     "name": "Alice",[0m
[33m!     "name": "Alice Johnson",[0m
      "email": "alice@example.com",[0m
[31m-     "password": "*****",[0m
[33m!     "password": "*****",[0m
      "preferences": {[0m
[31m-          "theme": "dark",[0m
[33m!          "theme": "light",[0m
[31m-          "notifications": true,[0m
[33m!          "notifications": false,[0m
           "layout": [[0m
[31m-               0: "grid",[0m
[33m!               0: "list",[0m
[31m-               1: "compact",[0m
           ],[0m
[32m+          "newFeature": true,[0m
      },[0m
      "metadata": Map (4) {[0m
[31m-          {"complex":"key"}: "value",[0m
[33m!          {"complex":"key"}: "newValue",[0m
[31m-          "role": "admin",[0m
[33m!          "role": "user",[0m
           "permissions": Set [[0m
                "read",[0m
[31m-               "write",[0m
                "execute",[0m
           ],[0m
[32m+          "createdAt": 1709294400000,[0m
      },[0m
      "tags": Set [[0m
           "user",[0m
[31m-          "active",[0m
[32m+          "inactive",[0m
[32m+          "premium",[0m
      ],[0m
[31m-     "lastLogin": 1709294400000,[0m
[33m!     "lastLogin": 1711096200000,[0m
      "custom": {[0m
[31m-          "key": "value",[0m
[33m!          "key": "modified",[0m
      },[0m
      "emptyArray": [[0m
[32m+          0: "unexpected",[0m
      ],[0m
      "emptyObject": {[0m
[32m+          "newKey": "new",[0m
      },[0m
      "nestedEmpty": {[0m
           "a": {[0m
[32m+               "b": "value",[0m
           },[0m
      },[0m
      "self": {[0m
           "id": 123,[0m
[31m-          "name": "Alice",[0m
[33m!          "name": "Alice Johnson",[0m
           "email": "alice@example.com",[0m
[31m-          "password": "*****",[0m
[33m!          "password": "*****",[0m
           "preferences": {[0m
[31m-               "theme": "dark",[0m
[33m!               "theme": "light",[0m
[31m-               "notifications": true,[0m
[33m!               "notifications": false,[0m
                "layout": [[0m
[31m-                    0: "grid",[0m
[33m!                    0: "list",[0m
[31m-                    1: "compact",[0m
                ],[0m
[32m+               "newFeature": true,[0m
           },[0m
           "metadata": Map (4) {[0m
[31m-               {"complex":"key"}: "value",[0m
[33m!               {"complex":"key"}: "newValue",[0m
[31m-               "role": "admin",[0m
[33m!               "role": "user",[0m
                "permissions": Set [[0m
                     "read",[0m
[31m-                    "write",[0m
                     "execute",[0m
                ],[0m
[32m+               "createdAt": 1709294400000,[0m
           },[0m
           "tags": Set [[0m
                "user",[0m
[31m-               "active",[0m
[32m+               "inactive",[0m
[32m+               "premium",[0m
           ],[0m
[31m-          "lastLogin": 1709294400000,[0m
[33m!          "lastLogin": 1711096200000,[0m
           "custom": {[0m
[31m-               "key": "value",[0m
[33m!               "key": "modified",[0m
           },[0m
           "emptyArray": [[0m
[32m+               0: "unexpected",[0m
           ],[0m
           "emptyObject": {[0m
[32m+               "newKey": "new",[0m
           },[0m
           "nestedEmpty": {[0m
                "a": {[0m
[32m+                    "b": "value",[0m
                },[0m
           },[0m
           "self": [Circular],[0m
[32m+          "newProp": "New Value",[0m
      },[0m
[32m+     "newProp": "New Value",[0m
}[0m
\`\`\`"
`;

exports[`toDiffString outputs with defaults 1`] = `
"{[0m
   "id": 123,[0m
[31m-  "name": "Alice",[0m
[33m!  "name": "Alice Johnson",[0m
   "email": "alice@example.com",[0m
[31m-  "password": "*****",[0m
[33m!  "password": "*****",[0m
   "preferences": {[0m
[31m-    "theme": "dark",[0m
[33m!    "theme": "light",[0m
[31m-    "notifications": true,[0m
[33m!    "notifications": false,[0m
     "layout": [[0m
[31m-      0: "grid",[0m
[33m!      0: "list",[0m
[31m-      1: "compact",[0m
     ],[0m
[32m+    "newFeature": true,[0m
   },[0m
   "metadata": Map (4) {[0m
[31m-    {"complex":"key"}: "value",[0m
[33m!    {"complex":"key"}: "newValue",[0m
[31m-    "role": "admin",[0m
[33m!    "role": "user",[0m
     "permissions": Set [[0m
       "read",[0m
[31m-      "write",[0m
       "execute",[0m
     ],[0m
[32m+    "createdAt": 1709294400000,[0m
   },[0m
   "tags": Set [[0m
     "user",[0m
[31m-    "active",[0m
[32m+    "inactive",[0m
[32m+    "premium",[0m
   ],[0m
[31m-  "lastLogin": 1709294400000,[0m
[33m!  "lastLogin": 1711096200000,[0m
   "custom": {[0m
[31m-    "key": "value",[0m
[33m!    "key": "modified",[0m
   },[0m
   "emptyArray": [[0m
[32m+    0: "unexpected",[0m
   ],[0m
   "emptyObject": {[0m
[32m+    "newKey": "new",[0m
   },[0m
   "nestedEmpty": {[0m
     "a": {[0m
[32m+      "b": "value",[0m
     },[0m
   },[0m
   "self": {[0m
     "id": 123,[0m
[31m-    "name": "Alice",[0m
[33m!    "name": "Alice Johnson",[0m
     "email": "alice@example.com",[0m
[31m-    "password": "*****",[0m
[33m!    "password": "*****",[0m
     "preferences": {[0m
[31m-      "theme": "dark",[0m
[33m!      "theme": "light",[0m
[31m-      "notifications": true,[0m
[33m!      "notifications": false,[0m
       "layout": [[0m
[31m-        0: "grid",[0m
[33m!        0: "list",[0m
[31m-        1: "compact",[0m
       ],[0m
[32m+      "newFeature": true,[0m
     },[0m
     "metadata": Map (4) {[0m
[31m-      {"complex":"key"}: "value",[0m
[33m!      {"complex":"key"}: "newValue",[0m
[31m-      "role": "admin",[0m
[33m!      "role": "user",[0m
       "permissions": Set [[0m
         "read",[0m
[31m-        "write",[0m
         "execute",[0m
       ],[0m
[32m+      "createdAt": 1709294400000,[0m
     },[0m
     "tags": Set [[0m
       "user",[0m
[31m-      "active",[0m
[32m+      "inactive",[0m
[32m+      "premium",[0m
     ],[0m
[31m-    "lastLogin": 1709294400000,[0m
[33m!    "lastLogin": 1711096200000,[0m
     "custom": {[0m
[31m-      "key": "value",[0m
[33m!      "key": "modified",[0m
     },[0m
     "emptyArray": [[0m
[32m+      0: "unexpected",[0m
     ],[0m
     "emptyObject": {[0m
[32m+      "newKey": "new",[0m
     },[0m
     "nestedEmpty": {[0m
       "a": {[0m
[32m+        "b": "value",[0m
       },[0m
     },[0m
     "self": [Circular],[0m
[32m+    "newProp": "New Value",[0m
   },[0m
[32m+  "newProp": "New Value",[0m
}[0m"
`;

exports[`toDiffString outputs without color 1`] = `
"{
   "id": 123,
-  "name": "Alice",
!  "name": "Alice Johnson",
   "email": "alice@example.com",
-  "password": "*****",
!  "password": "*****",
   "preferences": {
-    "theme": "dark",
!    "theme": "light",
-    "notifications": true,
!    "notifications": false,
     "layout": [
-      0: "grid",
!      0: "list",
-      1: "compact",
     ],
+    "newFeature": true,
   },
   "metadata": Map (4) {
-    {"complex":"key"}: "value",
!    {"complex":"key"}: "newValue",
-    "role": "admin",
!    "role": "user",
     "permissions": Set [
       "read",
-      "write",
       "execute",
     ],
+    "createdAt": 1709294400000,
   },
   "tags": Set [
     "user",
-    "active",
+    "inactive",
+    "premium",
   ],
-  "lastLogin": 1709294400000,
!  "lastLogin": 1711096200000,
   "custom": {
-    "key": "value",
!    "key": "modified",
   },
   "emptyArray": [
+    0: "unexpected",
   ],
   "emptyObject": {
+    "newKey": "new",
   },
   "nestedEmpty": {
     "a": {
+      "b": "value",
     },
   },
   "self": {
     "id": 123,
-    "name": "Alice",
!    "name": "Alice Johnson",
     "email": "alice@example.com",
-    "password": "*****",
!    "password": "*****",
     "preferences": {
-      "theme": "dark",
!      "theme": "light",
-      "notifications": true,
!      "notifications": false,
       "layout": [
-        0: "grid",
!        0: "list",
-        1: "compact",
       ],
+      "newFeature": true,
     },
     "metadata": Map (4) {
-      {"complex":"key"}: "value",
!      {"complex":"key"}: "newValue",
-      "role": "admin",
!      "role": "user",
       "permissions": Set [
         "read",
-        "write",
         "execute",
       ],
+      "createdAt": 1709294400000,
     },
     "tags": Set [
       "user",
-      "active",
+      "inactive",
+      "premium",
     ],
-    "lastLogin": 1709294400000,
!    "lastLogin": 1711096200000,
     "custom": {
-      "key": "value",
!      "key": "modified",
     },
     "emptyArray": [
+      0: "unexpected",
     ],
     "emptyObject": {
+      "newKey": "new",
     },
     "nestedEmpty": {
       "a": {
+        "b": "value",
       },
     },
     "self": [Circular],
+    "newProp": "New Value",
   },
+  "newProp": "New Value",
}"
`;
